### PR TITLE
util: refactor splitter pt. 2 (auto-updater)

### PR DESCRIPTION
**Pull request by wexxlee**
Part 2 of the work on #94 (see #109 for more info).

~~This is built on the type-name changes in #127, so that should be merged first (and the changes in `94fd33e` can be ignored, as I will rebase them out before merging).~~

This also sets both `ActorControl` and `NameToggle` to `include: 'never'`.  Both are used ubiquitously in timelines (83 and 60 files, respectively) as wipe regex and to sync --untargetable-- and --targetable-- entries, and since those use cases are handled automatically by timeline tools, I wouldn't think they would have much utility via analysis filtering (although open to leaving them in if others feel differently?).  I'm mostly disabling them from the filter in this PR to avoid massively cluttering the auto-generated PR that will follow.  Speaking of which...

When this is merged, I expect an auto-generated PR to flag uses of (and attempt to add `include: 'all'` to):
  - `RemovedCombatant` (used in 3 triggers, 2 timelines)
  - `NetworkCancelAbility` (used in 2 timelines)
  - `WasDefeated` (used in 3 triggers, 1 timeline)
  - `NetworkEffectResult` (used in 1 trigger)
  - `StartsUsingExtra` (used in 2 triggers)

That said, I think we can save discussion & changes to those for the auto-generated PR.

### CODE REVIEW

**Comment by valarnin** [comment]
> This also sets both `ActorControl` and `NameToggle` to `include: 'never'`. Both are used ubiquitously in timelines (83 and 60 files, respectively) as wipe regex and to sync --untargetable-- and --targetable-- entries, and since those use cases are handled automatically by timeline tools, I wouldn't think they would have much utility via analysis filtering (although open to leaving them in if others feel differently?). I'm mostly disabling them from the filter in this PR to avoid massively cluttering the auto-generated PR that will follow. Speaking of which...

Any log lines that may be even tangentially related to triggers or timelines for a fight should be included, unless I'm misunderstanding what you're doing here. If a user uses the splitter to split a log file into individual encounters for playback in raidemulator, for example, all of the log lines that would be triggered on (including AC and NameToggle) should be included. This is also for purposes of e.g. testing timelines via the `test_timeline.ts` tool.

**Comment by wexxlee** [comment]
> Any log lines that may be even tangentially related to triggers or timelines for a fight should be included, unless I'm misunderstanding what you're doing here. If a user uses the splitter to split a log file into individual encounters for playback in raidemulator, for example, all of the log lines that would be triggered on (including AC and NameToggle) should be included. This is also for purposes of e.g. testing timelines via the `test_timeline.ts` tool.

I think there might be a misunderstanding.  All log lines are included by default whenever a log is split. The `analysisOptions` property in `netlog_defs` controls only whether (and how many) of those log lines are included if the 'Filter Log for Analysis' checkbox is used in the log splitter tool (or the corresponding CLI option).  The [intent of that filter option](https://github.com/OverlayPlugin/cactbot/pull/17) was to (mostly) mirror the original regex that quisquous used to reduce a log file (or encounter) down to "interesting" lines to make it easier to create timelines & triggers.  But that regex was based on ACT (not "network") lines, and it hadn't been updated to include a lot of the recently-added line types.  The filter option was an effort to update and integrate that regex directly into the splitter tool.

The filter will (and should, I think) exclude a fair number of lines that aren't likely to be of interest when analyzing a log file for that purpose.  I wouldn't expect the filter to be the default for most users, and especially not for logs being used for debugging/raid emulator/etc.  Its sole intent is to trim logs to make them more readable for those of us actively working on trigger/timeline development.  Perhaps that could be clearer in the UI?

**Comment by xiashtra** [comment]
> > Any log lines that may be even tangentially related to triggers or timelines for a fight should be included, unless I'm misunderstanding what you're doing here. If a user uses the splitter to split a log file into individual encounters for playback in raidemulator, for example, all of the log lines that would be triggered on (including AC and NameToggle) should be included. This is also for purposes of e.g. testing timelines via the `test_timeline.ts` tool.
> 
> I think there might be a misunderstanding. All log lines are included by default whenever a log is split. The `analysisOptions` property in `netlog_defs` controls only whether (and how many) of those log lines are included if the 'Filter Log for Analysis' checkbox is used in the log splitter tool (or the corresponding CLI option). The [intent of that filter option](https://github.com/OverlayPlugin/cactbot/pull/17) was to (mostly) mirror the original regex that quisquous used to reduce a log file (or encounter) down to "interesting" lines to make it easier to create timelines & triggers. But that regex was based on ACT (not "network") lines, and it hadn't been updated to include a lot of the recently-added line types. The filter option was an effort to update and integrate that regex directly into the splitter tool.
> 
> The filter will (and should, I think) exclude a fair number of lines that aren't likely to be of interest when analyzing a log file for that purpose. I wouldn't expect the filter to be the default for most users, and especially not for logs being used for debugging/raid emulator/etc. Its sole intent is to trim logs to make them more readable for those of us actively working on trigger/timeline development. Perhaps that could be clearer in the UI?

As long as the UI is clear to an end user which options are meant for splitting a log to send to us for debugging vs which options are meant more for developers I think it should be ok.

File: util/update_logdefs.ts
Comment (valarnin) [inline util/update_logdefs.ts:181]: As always, I'm going to grumble here about parsing files with regex when another option is available.

Would it be possible to do something like:

1. Import the file normally
2. Loop through the `triggers` and `timelineTriggers`
3. If there's a trigger found, then regex the file for something like `id: '${trigger.id}'.*?(?:netRegex|regex): .*?$` to find the line number

File: util/update_logdefs.ts
Comment (valarnin) [inline util/update_logdefs.ts:241]: Could we not pull the line's decimal type from the constructed regex (via `^\^\(\?<type>([^)]+)\)`, remember to account for `Ability` being `2[12]`), and therefore leverage the already existing timeline parsing code?

We could then match this back to an actual def type trivially in log defs.

**Review by valarnin — COMMENTED** [review]
Sorry to keep complaining about parsing files with regex, but every time I see new regex to parse a file that's already being handled otherwise elsewhere, from my perspective it's another potential break point that can't easily be detected if there's any changes to the file format or other refactoring in the future.

Also, this.

https://stackoverflow.com/a/1732454

File: util/update_logdefs.ts
Comment (wexxlee) [reply util/update_logdefs.ts:181]: Yeah, I think that works.  I'd normally grumble about the cost of having to import and then loop-regex the file for each match, but the matches should be very small going forward, and I hate using regex to parse JS files (maybe not as much as you, but probably close).  Also, thanks -- that StackOverflow comment absolutely made my day. 🤣

File: util/update_logdefs.ts
Comment (wexxlee) [reply util/update_logdefs.ts:241]: Eesh.  It's a pretty creative solve, but I'm not wild about parsing regex with regex.  Also, I think your original concern -- about changes to the file format causing undetected breakage here -- would equally apply to regex'ing the timeline regex. But it did get me thinking about a potential alternative I want to run by you:

What if we add a no-op method to the TimelineParser class and call it at the right point during timeline parsing -- and then in this file, we do an extends class that overrides the method and extracts the type?  There would be negligible-to-no impact on runtime performance, I think, and a comment on the no-op method in the base class would help avoid possible breakage situations in the future.  Thoughts?

Comment (valarnin) [reply util/update_logdefs.ts:241]: > What if we add a no-op method to the TimelineParser class and call it at the right point during timeline parsing -- and then in this file, we do an extends class that overrides the method and extracts the type?

We do something similar for RaidEmulator at various points in the timeline and popup-text files, so I think this would be fine.

**Comment by wexxlee** [comment]
OK, this updates the script to remove the regex dependency for trigger and timeline logic per earlier discussion.  Separate commits for each to help with reviewability.  With this change, it also became fairly trivial to add oopsy trigger checking as well, so I added that in `dc14878`.

(Note that by adding oopsy, the auto-generated PR that will follow after merging this will now propose to add NetworkDoT to the analysis filter as well.)

File: ui/raidboss/timeline_parser.ts
Comment (valarnin) [inline ui/raidboss/timeline_parser.ts:189]: Could this exist on the subtype instead, and the type assertion be changed to account for it? e.g.
```ts
type TimelineParserWithEntries = TimelineParser & {
  entries: Partial<Record<LogDefinitionName, number[]>> | undefined;
};
// ...
    const triggerSet = (await import(importPath)).default as TimelineParserWithEntries;
```

(feel free to name the type something else)

Or you could put the `entries` property on the subclass directly and just `.default as TimelineTypeExtractor`?

File: util/update_logdefs.ts
Comment (valarnin) [inline util/update_logdefs.ts:22]: This comment still mentions `'none'` which isn't valid anymore.

File: util/update_logdefs.ts
Comment (valarnin) [inline util/update_logdefs.ts:366]: I'm still not really happy with how this function works in general, even after a second review, but I can't really see a better way to handle it.

My instincts are telling me that it should be possible to do something like:
1. Import the current `latestLogDefinitions`
2. Clone the object, stripping out the `const`
3. Update the cloned object to have the new properties
4. `JSON.stringify` it
5. Replace the contents in the file

But I can't really figure out the process in more detail than that right now, so I'm not sure if it would actually work.

Consider this a general comment/thought rather than a request for change.

**Review by valarnin — CHANGES_REQUESTED** [review]

File: ui/raidboss/timeline_parser.ts
Comment (wexxlee) [reply ui/raidboss/timeline_parser.ts:189]: I don't think this would quite work, as timelines aren't imported - `TimelineParser` takes a string input (file read with newlines) and the constructor calls `parse()` to process the string.  Or am I misunderstanding your suggestion?

Taking a step back though, the reason I added `entries` as a property on the parent instead of the subclass was because the parent constructor calls `parse()` (which starts the method chain that ultimately results in `entries` being populated).  So defining it on the subclass results in re-initialization that wipes the data, as subclass props are initialized after parent class construction.  It was a bit too tortured to try to handle type extraction as its own callable method outside of the `parse()` chain, because all of the regex/entry processing is done by that point, and I didn't want to change `parse()` to be a callable method, since that would require reworking all current implementations of `TimelineParser`.  I wasn't crazy about making `entries` a parent prop, but it was the simplest solution and at least made logical sense, since the parent class is technically populating it.

File: util/update_logdefs.ts
Comment (wexxlee) [reply util/update_logdefs.ts:366]: I share the motivation, and I think that conceptually works, but the problem is that it will blow away any inline comments in the const, as they won't be imported.  I couldn't figure out a way around that wouldn't involve regex'ing all comments and then trying to add them back in after `JSON.stringify`, and that would be even more gross...

File: ui/raidboss/timeline_parser.ts
Comment (valarnin) [reply ui/raidboss/timeline_parser.ts:189]: This diff works fine for me, no errors running the npm script and makes the expected changes to the netlog defs ts file.

```patch
diff --git a/ui/raidboss/timeline_parser.ts b/ui/raidboss/timeline_parser.ts
index 23b08d18b..0375f0658 100644
--- a/ui/raidboss/timeline_parser.ts
+++ b/ui/raidboss/timeline_parser.ts
@@ -185,8 +185,6 @@ export class TimelineParser {
   private labelToTime: { [name: string]: number } = {};
   // Map of encountered syncs to the label they are jumping to.
   private labelToSync: { [name: string]: Sync[] } = {};
-  // Used in subclass (update_logdefs), but defined here due to being utilized during construction
-  public entries: Partial<Record<LogDefinitionName, number[]>> | undefined;
 
   constructor(
     text: string,
diff --git a/util/update_logdefs.ts b/util/update_logdefs.ts
index d3ef2ed36..5437619b3 100644
--- a/util/update_logdefs.ts
+++ b/util/update_logdefs.ts
@@ -47,6 +47,8 @@ type FileMatch = {
 type FileMatches = Partial<Record<LogDefinitionName, FileMatch[]>>;
 
 class TimelineTypeExtractor extends TimelineParser {
+  public entries: Partial<Record<LogDefinitionName, number[]>> = {};
+
   constructor(contents: string) {
     super(contents, [], []);
   }

```

Comment (wexxlee) [reply ui/raidboss/timeline_parser.ts:189]: > This diff works fine for me, no errors running the npm script and makes the expected changes to the netlog defs ts file.

It makes the same changes to `netlog_defs` right now because all of the new log types being added to the analysis filter can be and are being picked up from trigger file parsing; the timeline parsing has some log-type overlap but isn't adding any unique log types by itself, so comparing `netlog_defs` won't give the full picture of whether timeline parsing is working correctly.  OTOH, if you compare the script console outputs from having `entries` on the parent class vs. subclass, you'll see that having it on the subclass isn't picking up any of the uses of these log types in timeline files (see attached output).  (fwiw, i confirmed this with some debug console logging that showed that `entries` was empty after the subclass completed initialization).
[_parentprop_stdout.txt](https://github.com/OverlayPlugin/cactbot/files/15102914/_parentprop_stdout.txt)
[_subclassprop_stdout.txt](https://github.com/OverlayPlugin/cactbot/files/15102916/_subclassprop_stdout.txt)

Comment (valarnin) [reply ui/raidboss/timeline_parser.ts:189]: Okay, I see the confusion now. I think for now, update the comment to include a `TODO` to revisit after refactoring the timeline parser code, and later on we can make `parse` public and move responsibility for calling `parse` up to the code that's constructing the `TimelineParser`?

**Review by valarnin — APPROVED** [review]
Tentative pre-approval pending the discussion in the previous review about updating the comment.

File: ui/raidboss/timeline_parser.ts
Comment (wexxlee) [reply ui/raidboss/timeline_parser.ts:189]: > Okay, I see the confusion now. I think for now, update the comment to include a `TODO` to revisit after refactoring the timeline parser code, and later on we can make `parse` public and move responsibility for calling `parse` up to the code that's constructing the `TimelineParser`?

Actually, I think I might've figured out a way to do this that will move the prop to the subclass.  I can add an optional boolean param to `TimelineParser` that controls the execution of `parse()` in the constructor, make `parse()` public, and then have the subclass initialize the parent without calling `parse()`, and just manually call it in the update script after the subclass is fully initialized.  Not quite as clean, but it shouldn't affect any current implementations.  Although I don't know if making `parse()` public in and of itself will cause any runtime degradation?

**Comment by wexxlee** [comment]
> Tentative pre-approval pending the discussion in the previous review about updating the comment.

Going to hold on merging this in case you want to take another look, since my last commit is a different approach than we talked about.

File: ui/raidboss/timeline_parser.ts
Comment (wexxlee) [reply ui/raidboss/timeline_parser.ts]: > Why not just call `parse` directly as part of the subclass's constructor? e.g.

Hey, even better.

**Comment by valarnin** [comment]
Looks good to me now.

**Pull request merged**
